File: src/app/chunk.ts
import vec3 from 'gl-vec3';
import ndarray from 'ndarray';
import { BufferAttribute, BufferGeometry, Mesh } from 'three';

import { Engine } from '..';
import { Coords2, Coords3 } from '../libs';
import { simpleCull } from '../libs/meshers';
import { makeHeightMap } from '../libs/meshers/make-height-map';
import { Helper } from '../utils';

type ChunkOptions = {
  size: number;
  maxHeight: number;
  dimension: number;
  padding: number;
};

class Chunk {
  public engine: Engine;
  public coords: Coords2;
  public voxels: ndarray;
  public lights: ndarray;
  public heightMap: ndarray;

  public name: string;
  public size: number;
  public maxHeight: number;
  public dimension: number;
  public padding: number;
  public width: number;

  // voxel position references in voxel space
  public minInner: Coords3; // chunk's minimum voxel (not padded)
  public minOuter: Coords3; // chunk's minimum voxel (padded)
  public maxInner: Coords3; // chunk's maximum voxel (not padded)
  public maxOuter: Coords3; // chunk's maximum voxel (padded)

  public geometry: BufferGeometry;
  public mesh: Mesh;
  public altMesh: Mesh | undefined; // this way, this.mesh is in existence until ready to change, avoiding empty chunk frames

  public isEmpty = true;
  public isDirty = true;
  public isAdded = false;
  public isMeshing = false; // is meshing
  public isInitialized = false; // is populated with terrain info
  public isPending = false; // pending for client-side terrain generation
  public sunlightApplied = false;

  constructor(engine: Engine, coords: Coords2, { size, dimension, padding, maxHeight }: ChunkOptions) {
    this.engine = engine;
    this.coords = coords;

    this.size = size;
    this.maxHeight = maxHeight;
    this.dimension = dimension;
    this.padding = padding;
    this.width = size + padding * 2;
    this.name = Helper.getChunkName(this.coords);

    this.voxels = ndarray(new Int8Array(this.width * this.maxHeight * this.width), [
      this.width,
      this.maxHeight,
      this.width,
    ]);
    this.lights = ndarray(new Int8Array(this.width * this.maxHeight * this.width), [
      this.width,
      this.maxHeight,
      this.width,
    ]);
    this.heightMap = ndarray(new Int8Array(this.width * this.width), [this.width, this.width]);

    this.geometry = new BufferGeometry();

    this.minInner = [0, 0, 0];
    this.minOuter = [0, 0, 0];
    this.maxInner = [0, 0, 0];
    this.maxOuter = [0, 0, 0];

    const [cx, cz] = coords;
    const coords3 = [cx, 0, cz];

    // initialize
    vec3.copy(this.minInner, coords3);
    vec3.copy(this.minOuter, coords3);
    vec3.copy(this.maxInner, coords3);
    vec3.copy(this.maxOuter, coords3);

    // calculate
    const paddingVec = [padding, 0, padding];
    vec3.scale(this.minOuter, this.minOuter, size);
    vec3.sub(this.minOuter, this.minOuter, paddingVec);
    vec3.add(this.minInner, this.minOuter, paddingVec);
    vec3.add(this.maxOuter, this.maxOuter, [1, 0, 1]);
    vec3.scale(this.maxOuter, this.maxOuter, size);
    vec3.add(this.maxOuter, this.maxOuter, [0, maxHeight, 0]);
    vec3.add(this.maxOuter, this.maxOuter, paddingVec);
    vec3.sub(this.maxInner, this.maxOuter, paddingVec);
  }

  // goes from [-padding, 0, -padding] to [size + padding - 1, maxHeight - 1, size + padding - 1]
  getLocal(lx: number, ly: number, lz: number) {
    return this.voxels.get(lx + this.padding, ly, lz + this.padding);
  }

  // goes from [-padding, 0, -padding] to [size + padding - 1, maxHeight - 1, size + padding - 1]
  setLocal(lx: number, ly: number, lz: number, id: number) {
    return this.voxels.set(lx + this.padding, ly, lz + this.padding, id);
  }

  // goes from [-padding, 0, -padding] to [size + padding - 1, maxHeight - 1, size + padding - 1]
  getLocalTorchLight(lx: number, ly: number, lz: number) {
    return this.lights.get(lx + this.padding, ly, lz + this.padding) & 0xf;
  }

  // goes from [-padding, 0, -padding] to [size + padding - 1, maxHeight - 1, size + padding - 1]
  setLocalTorchLight(lx: number, ly: number, lz: number, level: number) {
    return this.lights.set(
      lx + this.padding,
      ly,
      lz + this.padding,
      (this.getLocalTorchLight(lx, ly, lz) & 0xf0) | level,
    );
  }

  // goes from [-padding, 0, -padding] to [size + padding - 1, maxHeight - 1, size + padding - 1]
  getLocalSunlight(lx: number, ly: number, lz: number) {
    return (this.lights.get(lx + this.padding, ly, lz + this.padding) >> 4) & 0xf;
  }

  // goes from [-padding, 0, -padding] to [size + padding - 1, maxHeight - 1, size + padding - 1]
  setLocalSunlight(lx: number, ly: number, lz: number, level: number) {
    return this.lights.set(
      lx + this.padding,
      ly,
      lz + this.padding,
      (this.getLocalSunlight(lx, ly, lz) & 0xf) | (level << 4),
    );
  }

  getVoxel(vx: number, vy: number, vz: number) {
    if (!this.contains(vx, vy, vz)) return;
    const [lx, ly, lz] = this.toLocal(vx, vy, vz);
    return this.getLocal(lx, ly, lz);
  }

  setVoxel(vx: number, vy: number, vz: number, id: number) {
    if (!this.contains(vx, vy, vz)) return;
    // if voxel type doesn't change (might conflict with lighting)
    if (this.getVoxel(vx, vy, vz) === id) return;

    const [lx, ly, lz] = this.toLocal(vx, vy, vz);
    this.setLocal(lx, ly, lz, id);

    // change chunk state
    if (id !== 0) this.isEmpty = false;
    // mark chunk as dirty
    this.isDirty = true;
  }

  getTorchLight(vx: number, vy: number, vz: number) {
    // if (!this.contains(vx, vy, vz)) return 0; // ?
    const [lx, ly, lz] = this.toLocal(vx, vy, vz);
    return this.getLocalTorchLight(lx, ly, lz);
  }

  setTorchLight(vx: number, vy: number, vz: number, level: number) {
    if (!this.contains(vx, vy, vz)) return;
    const [lx, ly, lz] = this.toLocal(vx, vy, vz);
    this.setLocalTorchLight(lx, ly, lz, level);
    this.isDirty = true; // mesh rebuilt needed if light changes
  }

  getSunlight(vx: number, vy: number, vz: number) {
    const [lx, ly, lz] = this.toLocal(vx, vy, vz);
    return this.getLocalSunlight(lx, ly, lz);
  }

  setSunlight(vx: number, vy: number, vz: number, level: number) {
    if (!this.contains(vx, vy, vz)) return;
    const [lx, ly, lz] = this.toLocal(vx, vy, vz);
    this.setLocalSunlight(lx, ly, lz, level);
    this.isDirty = true;
  }

  contains(vx: number, vy: number, vz: number, padding = this.padding) {
    const { size, maxHeight } = this;
    const [lx, ly, lz] = this.toLocal(vx, vy, vz);

    return lx >= -padding && lx < size + padding && ly >= 0 && ly < maxHeight && lz >= -padding && lz < size + padding;
  }

  distTo(vx: number, _: number, vz: number) {
    const [mx, , mz] = this.minInner;
    return Math.sqrt((mx - vx) * (mx - vx) + (mz - vz) * (mz - vz));
  }

  addToScene() {
    this.removeFromScene();
    if (!this.isAdded && this.altMesh) {
      this.engine.rendering.scene.add(this.altMesh);
      this.engine.world.addAsVisible(this);
      this.mesh = this.altMesh;
      this.isAdded = true;
    }
  }

  removeFromScene() {
    if (this.isAdded && this.mesh) {
      this.engine.rendering.scene.remove(this.mesh);
      this.engine.world.removeAsVisible(this);
      this.isAdded = false;
    }
  }

  async initialized() {
    this.isInitialized = true;

    // build mesh once initialized
    await makeHeightMap(this);
    // await this.engine.world.applySunlight(this);
    await this.buildMesh();
  }

  async buildMesh() {
    if (this.isEmpty) {
      // if it's empty, it can't be dirty
      this.isDirty = false;
      return;
    }

    // don't need to be meshed again
    this.isDirty = false;
    this.isMeshing = true;

    const { positions, normals, indices, uvs, aos, sunlights, torchLights } = await simpleCull(this);

    const positionNumComponents = 3;
    const normalNumComponents = 3;
    const uvNumComponents = 2;
    const occlusionNumComponents = 1;
    const sunlightsNumComponents = 1;
    const torchLightsNumComponents = 1;

    this.geometry.dispose();
    this.geometry.setAttribute('position', new BufferAttribute(positions, positionNumComponents));
    this.geometry.setAttribute('normal', new BufferAttribute(normals, normalNumComponents));
    this.geometry.setAttribute('uv', new BufferAttribute(uvs, uvNumComponents));
    this.geometry.setAttribute('ao', new BufferAttribute(aos, occlusionNumComponents));
    this.geometry.setAttribute('sunlight', new BufferAttribute(sunlights, sunlightsNumComponents));
    this.geometry.setAttribute('torchLight', new BufferAttribute(torchLights, torchLightsNumComponents));
    this.geometry.setIndex(Array.from(indices));

    this.altMesh = new Mesh(this.geometry, this.engine.registry.material);
    this.altMesh.name = this.name;
    this.altMesh.renderOrder = 10000;
    this.altMesh.frustumCulled = false;

    // mark chunk as built mesh
    this.isMeshing = false;
    this.isPending = false;
  }

  private toLocal = (vx: number, vy: number, vz: number) => {
    return vec3.sub([0, 0, 0], [vx, vy, vz], this.minInner);
  };
}

export { Chunk };

File: server-rs/core/chunk.rs
use ndarray::{Array2, Array3};

use crate::{
    libs::types::{Coords2, Coords3},
    utils::convert,
};

#[derive(Debug)]
pub struct Chunk {
    pub name: String,

    pub coords: Coords2<i32>,

    pub voxels: Array3<u8>,
    pub lights: Array3<u8>,
    pub height_map: Array2<i32>,

    pub min: Coords3<i32>,
    pub max: Coords3<i32>,

    pub needs_saving: bool,
    pub needs_propagation: bool,
    pub needs_terrain: bool,
    pub needs_decoration: bool,

    pub is_empty: bool,
    pub is_dirty: bool,

    pub top_y: i32,

    pub size: usize,
    pub max_height: usize,
}

impl Chunk {
    pub fn new(coords: Coords2<i32>, size: usize, max_height: usize) -> Self {
        let Coords2(cx, cz) = coords;

        let name = convert::get_chunk_name(&coords);

        let voxels: Array3<u8> = Array3::zeros((size, max_height, size));
        let lights: Array3<u8> = Array3::zeros((size, max_height, size));
        let height_map: Array2<i32> = Array2::zeros((size, size));

        let coords3 = Coords3(cx, 0, cz);

        let min = coords3.clone().scale(size as i32);
        let max = coords3
            .clone()
            .add(&Coords3(1, 0, 1))
            .scale(size as i32)
            .add(&Coords3(0, max_height as i32, 0));

        Self {
            name,

            coords,
            voxels,
            lights,
            height_map,
            min,
            max,

            needs_saving: false,
            needs_propagation: true,
            needs_terrain: true,
            needs_decoration: true,

            is_empty: true,
            is_dirty: true,

            top_y: max_height as i32,

            size,
            max_height,
        }
    }

    pub fn get_voxel(&self, vx: i32, vy: i32, vz: i32) -> u8 {
        if !self.contains(vx, vy, vz, 0) {
            return 0;
        }

        let Coords3(lx, ly, lz) = self.to_local(vx, vy, vz);
        self.voxels[[lx as usize, ly as usize, lz as usize]]
    }

    pub fn set_voxel(&mut self, vx: i32, vy: i32, vz: i32, id: u8) {
        assert!(self.contains(vx, vy, vz, 0));

        let Coords3(lx, ly, lz) = self.to_local(vx, vy, vz);
        self.voxels[[lx as usize, ly as usize, lz as usize]] = id;
    }

    pub fn get_torch_light(&self, vx: i32, vy: i32, vz: i32) -> u8 {
        if !self.contains(vx, vy, vz, 0) {
            return 0;
        }

        let Coords3(lx, ly, lz) = self.to_local(vx, vy, vz);
        self.get_local_torch_light(lx as usize, ly as usize, lz as usize)
    }

    pub fn set_torch_light(&mut self, vx: i32, vy: i32, vz: i32, level: u8) {
        assert!(self.contains(vx, vy, vz, 0));

        let Coords3(lx, ly, lz) = self.to_local(vx, vy, vz);
        self.set_local_torch_light(lx as usize, ly as usize, lz as usize, level)
    }

    pub fn get_sunlight(&self, vx: i32, vy: i32, vz: i32) -> u8 {
        if !self.contains(vx, vy, vz, 0) {
            return 0;
        }

        let Coords3(lx, ly, lz) = self.to_local(vx, vy, vz);
        self.get_local_sunlight(lx as usize, ly as usize, lz as usize)
    }

    pub fn set_sunlight(&mut self, vx: i32, vy: i32, vz: i32, level: u8) {
        assert!(self.contains(vx, vy, vz, 0));

        let Coords3(lx, ly, lz) = self.to_local(vx, vy, vz);
        self.set_local_sunlight(lx as usize, ly as usize, lz as usize, level)
    }

    pub fn get_max_height(&self, column: &Coords2<i32>) -> i32 {
        if !self.contains(column.0, 0, column.1, 0) {
            return self.max_height as i32;
        }

        let Coords3(lx, _, lz) = self.to_local(column.0, 0, column.1);
        self.height_map[[lx as usize, lz as usize]]
    }

    pub fn set_max_height(&mut self, column: &Coords2<i32>, height: i32) {
        assert!(self.contains(column.0, 0, column.1, 0));

        if height > self.top_y {
            self.top_y = height;
        }

        let Coords3(lx, _, lz) = self.to_local(column.0, 0, column.1);
        self.height_map[[lx as usize, lz as usize]] = height;
    }

    pub fn load() {
        todo!()
    }

    pub fn save() {
        todo!()
    }

    pub fn generate() {
        todo!()
    }

    pub fn decorate() {
        todo!()
    }

    pub fn generate_height_map() {
        todo!()
    }

    pub fn propagate() {
        todo!()
    }

    pub fn update() {
        todo!()
    }

    pub fn remesh() {
        todo!()
    }

    pub fn get_protocol() {
        todo!()
    }

    fn get_local_torch_light(&self, lx: usize, ly: usize, lz: usize) -> u8 {
        self.lights[[lx, ly, lz]] & 0xf
    }

    fn set_local_torch_light(&mut self, lx: usize, ly: usize, lz: usize, level: u8) {
        self.lights[[lx, ly, lz]] = (self.lights[[lx, ly, lz]] & 0xf0) | level;
    }

    fn get_local_sunlight(&self, lx: usize, ly: usize, lz: usize) -> u8 {
        (self.lights[[lx, ly, lz]] >> 4) & 0xf
    }

    fn set_local_sunlight(&mut self, lx: usize, ly: usize, lz: usize, level: u8) {
        self.lights[[lx, ly, lz]] = (self.lights[[lx, ly, lz]] & 0xf) | (level << 4);
    }

    fn to_local(&self, vx: i32, vy: i32, vz: i32) -> Coords3<i32> {
        Coords3(vx, vy, vz).sub(&self.min)
    }

    fn flood_light() {
        todo!()
    }

    fn remove_light() {
        todo!()
    }

    fn contains(&self, vx: i32, vy: i32, vz: i32, padding: i32) -> bool {
        let size = self.size as i32;
        let max_height = self.max_height as i32;

        let Coords3(lx, ly, lz) = self.to_local(vx, vy, vz);

        lx >= -padding
            && lx < size + padding
            && ly >= 0
            && ly < max_height
            && lz >= -padding
            && lz < size + padding
    }
}
